src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc

```diff
@@ -0,0 +1,571 @@
+      });
+
+    _metastore.add_objects(std::move(builder), term_map).get().value();
+
+    auto reader = make_reader(ntp, tidp);
+    auto result = read_all(reader.get());
+    EXPECT_TRUE(result.empty());
+    EXPECT_TRUE(reader->is_end_of_stream());
+}
+
+TEST_F(l1_reader_test, empty_offset_range) {
+    auto [ntp, tidp] = make_ntidp("test_topic");
+
+    // Write some initial objects.
+    auto batches
+      = model::test::make_random_batches(model::offset{0}, 250).get();
+
+    for (auto start = 0; start <= 200; start += 50) {
+        std::vector<tidp_batches_t> tidp_batches;
+        auto subbatches = slice(batches, start, 50);
+        tidp_batches.emplace_back(tidp, std::move(subbatches));
+        make_l1_objects(tidp_batches);
+    }
+
+    // Mimic an object from which all partition data has been compacted.
+    // The object has no data for the partition, but is registered
+    // to cover a non-empty offset range in the metastore.
+    auto meta_builder = _metastore.object_builder();
+
+    auto oid = meta_builder->get_or_create_object_for(tidp);
+
+    auto buf = iobuf{};
+    auto builder = l1::object_builder::create(
+      make_iobuf_ref_output_stream(buf), {});
+
+    auto obj_info = builder->finish().get();
+    builder->close().get();
+
+    _io.put_object(oid, std::move(buf));
+
+    auto missing = kafka::next_offset(
+      model::offset_cast(batches.back().last_offset()));
+    meta_builder
+      ->add(
+        oid,
+        l1::metastore::object_metadata::ntp_metadata{
+          .tidp = tidp,
+          .base_offset = missing,
+          .last_offset = missing,
+          .max_timestamp = model::timestamp::now(),
+          .pos = 0,
+          .size = 0,
+        })
+      .value();
+
+    meta_builder->finish(oid, obj_info.footer_offset, obj_info.size_bytes)
+      .value();
+
+    l1::metastore::term_offset_map_t term_map;
+    term_map[tidp].push_back(
+      l1::metastore::term_offset{
+        .term = model::term_id{1},
+        .first_offset = missing,
+      });
+    _metastore.add_objects(std::move(meta_builder), term_map).get().value();
+
+    // Write some objects after the empty object.
+    auto final_batches = model::test::make_random_batches(
+                           kafka::offset_cast(kafka::next_offset(missing)), 250)
+                           .get();
+
+    for (auto start = 0; start <= 200; start += 50) {
+        std::vector<tidp_batches_t> tidp_batches;
+        auto subbatches = slice(final_batches, start, 50);
+        tidp_batches.emplace_back(tidp, std::move(subbatches));
+        make_l1_objects(tidp_batches);
+    }
+
+    auto reader = make_reader(ntp, tidp);
+    auto result = read_all(reader.get());
+
+    batches.insert(
+      batches.end(),
+      std::make_move_iterator(final_batches.begin()),
+      std::make_move_iterator(final_batches.end()));
+    EXPECT_EQ(result, batches);
+    EXPECT_TRUE(reader->is_end_of_stream());
+}
+
+TEST_F(l1_reader_test, sparse_offset_ranges) {
```

Comment:
Not needed right now, but this kind of thing is also useful to have randomized testing that creates a random amount of data with randomized boundaries and at various different offset boundaries matches what the reader returns.